File: src/main/GO/leetcode/lists/offer/Offer12.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2022/12/28 21:49
 */

func exist(board [][]byte, word string) bool {

	if len(board) == 0 {
		return false
	}

	visited := make([][]bool, len(board))
	for i := 0; i < len(board); i++ {
		visited[i] = make([]bool, len(board[0]))
	}

	var dfs func(board [][]byte, word string, index int, i int, j int) bool
	dfs = func(board [][]byte, word string, index int, i int, j int) bool {
		if i < 0 || j < 0 || i >= len(board) || j >= len(board[0]) || visited[i][j] ||
			index >= len(word) || board[i][j] != word[index] {
			return false
		}
		//fix
		if index == len(word)-1 {
			return true
		}
		visited[i][j] = true
		flag := dfs(board, word, index+1, i+1, j) ||
			dfs(board, word, index+1, i, j+1) ||
			dfs(board, word, index+1, i-1, j) ||
			dfs(board, word, index+1, i, j-1)
		visited[i][j] = false
		return flag
	}

	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			if board[i][j] == word[0] {
				res := dfs(board, word, 0, i, j)
				if res {
					return true
				}
			}
		}
	}
	return false
}

File: java/src/leetcode/lists/offer/Offer12.java
package leetcode.lists.offer;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/12 14:56
 */
public class Offer12 {

    char[][] board;
    boolean[][] visited;
    String word;

    public boolean exist(char[][] board, String word) {
        if (board == null || board.length == 0) {
            return false;
        }
        this.board = board;
        this.word = word;

        visited = new boolean[board.length][board[0].length];
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] == word.charAt(0)) {
                    if (dfs(i, j, 0)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private boolean dfs(int i, int j, int index) {
        if (i < 0 || j < 0 || i >= board.length || j >= board[0].length || board[i][j] != word.charAt(index) || visited[i][j]) {
            return false;
        }
        if (index == word.length() - 1) {
            return true;
        }
        visited[i][j] = true;
        boolean res = dfs(i + 1, j, index + 1) ||
                dfs(i - 1, j, index + 1) ||
                dfs(i, j + 1, index + 1) ||
                dfs(i, j - 1, index + 1);
        visited[i][j] = false;
        return res;
    }
}
